src/main/java/com/powsybl/ws/commons/computation/service/AbstractWorkerService.java

```diff
@@ -0,0 +1,258 @@
+        } finally {
+            lockRunAndCancel.unlock();
+        }
+    }
+
+    protected abstract AbstractResultContext<R> fromMessage(Message<String> message);
+
+    protected boolean resultCanBeSaved(S result) {
+        return result != null;
+    }
+
+    public Consumer<Message<String>> consumeRun() {
+        return message -> {
+            AbstractResultContext<R> resultContext = fromMessage(message);
+            try {
+                long startTime = System.nanoTime();
+
+                Network network = getNetwork(resultContext.getRunContext().getNetworkUuid(),
+                        resultContext.getRunContext().getVariantId());
+                resultContext.getRunContext().setNetwork(network);
+                S result = run(resultContext.getRunContext(), resultContext.getResultUuid());
+
+                LOGGER.info("Just run in {}s", TimeUnit.NANOSECONDS.toSeconds(System.nanoTime() - startTime));
+
+                if (resultCanBeSaved(result)) {
+                    startTime = System.nanoTime();
+                    observer.observe("results.save", resultContext.getRunContext(), () -> saveResult(network, resultContext, result));
+
+                    LOGGER.info("Stored in {}s", TimeUnit.NANOSECONDS.toSeconds(System.nanoTime() - startTime));
+
+                    sendResultMessage(resultContext, result);
+                    LOGGER.info("{} complete (resultUuid='{}')", getComputationType(), resultContext.getResultUuid());
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } catch (Exception e) {
+                if (!(e instanceof CancellationException)) {
+                    LOGGER.error(NotificationService.getFailedMessage(getComputationType()), e);
+                    publishFail(resultContext, e.getMessage());
+                    resultService.delete(resultContext.getResultUuid());
+                    this.handleNonCancellationException(resultContext, e);
+                }
+            } finally {
+                futures.remove(resultContext.getResultUuid());
+                cancelComputationRequests.remove(resultContext.getResultUuid());
+            }
+        };
+    }
+
+    /**
+     * Handle exception in consumeRun that is not a CancellationException
+     * @param resultContext The context of the computation
+     * @param exception The exception to handle
+     */
+    protected void handleNonCancellationException(AbstractResultContext<R> resultContext, Exception exception) {
+    }
+
+    public Consumer<Message<String>> consumeCancel() {
+        return message -> cancelAsync(CancelContext.fromMessage(message));
+    }
+
+    protected abstract void saveResult(Network network, AbstractResultContext<R> resultContext, S result);
+
+    protected void sendResultMessage(AbstractResultContext<R> resultContext, Map<String, Object> additionalHeaders) {
```

Comment:
These methodes sendResultMessage anf publishFail with additionalHeaders are not used, we can remove them

Reply:
Done.

Reply:
OK